update big model ram

psky1111 opened the pull request:
ram and ram plus pretrain：
链接：https://pan.baidu.com/s/1VV9PvEnC1e6yFj8Ccwr7Dg 
提取码：ifgr 
--来自百度网盘超级会员V6的分享

psky1111 committed:
update big model ram

psky1111 committed:
Update ramloss.py

psky1111 committed:
Update ram.py

cuicheng01 commented on deploy/python/postprocess.py:
```diff
@@ -496,3 +496,61 @@ def __call__(self, batch_preds, file_names=None):
+    
+class RamOutPut(object):
+
+    def __init__(self, language="cn",tag_list="", tag_list_chinese="", threshold=0.68, delete_tag_index=[], ram_class_threshold_path="ppcls/utils/RAM/ram_tag_list_threshold.txt"):
```
文件需要过一下pre-commit，ram_class_threshold_path在yaml中传入

cuicheng01 commented on deploy/python/predict_multimodal.py:
```diff
@@ -0,0 +1,52 @@
+import os
```
需要增加版权信息

cuicheng01 commented on ppcls/arch/backbone/clip/clip.py:
```diff
@@ -0,0 +1,513 @@
+import paddle
```
需要增加版权信息

cuicheng01 commented on ppcls/arch/backbone/clip/clip.py:
```diff
@@ -0,0 +1,513 @@
+    )
+    return model, get_transforms(224)
+
+def clip_vit_b_16_224():
```
这里的命名最好和paddleclas的整体命名保持一致

cuicheng01 commented on ppcls/arch/backbone/clip/tokenizer.py:
```diff
@@ -0,0 +1,140 @@
+import gzip
```
增加版权信息，增加代码引用信息

cuicheng01 commented on ppcls/configs/ram/config_bert.yaml:
```diff
@@ -0,0 +1,20 @@
+                'BertModel'
+              ]
+
+attention_probs_dropout_prob: 0.1
```
部分参数增加注释

cuicheng01 commented on ppcls/configs/ram/config_infer.yaml:
```diff
@@ -0,0 +1,37 @@
+  use_gpu: False
+  enable_mkldnn: False
+  cpu_num_threads: 10
+  enable_benchmark: True
```
这个应该是False

cuicheng01 commented on tools/export_model_multimodal.py:
```diff
@@ -0,0 +1,36 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
2023

cuicheng01 commented on ppcls/arch/backbone/legendary_models/swin_transformer.py:
```diff
@@ -444,7 +444,10 @@ def get_attn_mask(self, height, width, dtype):
         H, W = input_dimensions
         B, L, C = x.shape
-
+        if min(H, W) <= self.window_size:
```
改swin的目的是为了兼容ram吗？

cuicheng01 commented on ppcls/arch/backbone/ram/ram_plus.py:
```diff
@@ -0,0 +1,242 @@
+from paddle import nn
```
增加版权信息和代码引用信息

cuicheng01 commented on ppcls/arch/backbone/ram/ram_plus.py:
```diff
@@ -0,0 +1,242 @@
+    return text_embeddings
+
+
+CONFIG_PATH = 'ppcls'
```
不建议这么定死，应该在yaml中配置

cuicheng01 commented on ppcls/arch/backbone/ram/ram_plus.py:
```diff
@@ -0,0 +1,242 @@
+class RAM_plus(RAM):
+    def __init__(self,
+                 med_config=f'{CONFIG_PATH}/configs/ram/config_bert.yaml',
+                 image_size=384,
```
在train的yaml中配置

cuicheng01 commented on ppcls/arch/backbone/ram/ram_plus.py:
```diff
@@ -0,0 +1,242 @@
+                 threshold=0.68,
+                 delete_tag_index=[],
+                 pretrain_clip = '',
+                 tag_list=f'{CONFIG_PATH}/utils/ram/ram_tag_list.txt',
```
下边所有的yaml都应该在train的yaml中配置

cuicheng01 commented on ppcls/arch/backbone/ram/ram_plus.py:
```diff
@@ -0,0 +1,242 @@
+        return loss_tag, loss_dis, loss_alignment
+
+
+    def inference(self,
```
在模型中一般不定义inference，如定义，最好增加注释

cuicheng01 commented on ppcls/arch/backbone/ram/ram_plus.py:
```diff
@@ -0,0 +1,242 @@
+        label_embed_reweight = paddle.empty([bs, self.num_class, 512])
+
+        for i in range(bs):
+            # 这里对 value_ori 进行 reshape，然后使用 broadcasting
```
英文注释

cuicheng01 commented on ppcls/arch/backbone/ram/ram_plus.py:
```diff
@@ -0,0 +1,242 @@
+        logits = self.fc(tagging_embed[0]).squeeze(-1)
+
+
+        return logits, bs
```
注意代码规范

cuicheng01 commented on ppcls/configs/ram/config_infer.yaml:
```diff
@@ -0,0 +1,37 @@
+Global:
```
这个config应该放在deploy目录下

cuicheng01 commented on ppcls/configs/ram/config_q2l.yaml:
```diff
@@ -0,0 +1,21 @@
+                'BertModel'
+              ]
+
+attention_probs_dropout_prob: 0.1
```
增加参数解释

cuicheng01 commented on ppcls/configs/ram/config_train.yaml:
```diff
@@ -0,0 +1,111 @@
+# global configs
+Global:
+  checkpoints: null
+  pretrained_model: "ram_plus.pdparams"
```
需要规范命名，以及需要说明默认的pretrain是什么

cuicheng01 commented on ppcls/configs/ram/config_train.yaml:
```diff
@@ -0,0 +1,111 @@
+
+# mixed precision
+AMP:
+  use_amp: False
```
这里需要验证下amp是True的情况

cuicheng01 commented on ppcls/configs/ram/config_train.yaml:
```diff
@@ -0,0 +1,111 @@
+Arch:
+  name: ram_plus
+  vit: swin_l
+  pretrain_clip: ./ViT-B-32.pdparams
```
这里的参数名字建议换成clip_pretrained，另外建议这里简单注释一下，以免给不了解算法的用户造成困扰（即为什么swin加载的vit pretrain）

cuicheng01 commented on ppcls/configs/ram/config_train.yaml:
```diff
@@ -0,0 +1,111 @@
+  name: Momentum
+  momentum: 0.9
+  lr:
+    name: Piecewise
```
这里为什么不是Cosine？

cuicheng01 commented on ppcls/configs/ram/config_train.yaml:
```diff
@@ -0,0 +1,111 @@
+  Train:
+    - RAMLoss:
+        weight: 1.0
+        mode: "pretrain"
```
这里的mode也需要增加注释

cuicheng01 commented on ppcls/configs/ram/config_train.yaml:
```diff
@@ -0,0 +1,111 @@
+DataLoader:
+  Train:
+    dataset:
+      name: RAM_pretrain_dataset
```
dataset这里是一个类，注意类的命名规范

cuicheng01 commented on ppcls/configs/ram/config_train.yaml:
```diff
@@ -0,0 +1,111 @@
+
+    sampler:
+      name: DistributedBatchSampler
+      batch_size: 1
```
填写正确的bs

cuicheng01 commented on ppcls/configs/ram/config_train.yaml:
```diff
@@ -0,0 +1,111 @@
+      drop_last: False
+      shuffle: True
+    loader:
+      num_workers: 1
```
填写正确的num-worker

cuicheng01 commented on ppcls/data/dataloader/ram_dataset.py:
```diff
@@ -0,0 +1,150 @@
+import json
```
需要增加版权信息和代码引用

cuicheng01 commented on ppcls/data/dataloader/ram_dataset.py:
```diff
@@ -0,0 +1,150 @@
+from PIL import ImageFile
+ImageFile.LOAD_TRUNCATED_IMAGES = True
+Image.MAX_IMAGE_PIXELS = None
+
```
注意代码规范

cuicheng01 commented on ppcls/data/dataloader/ram_dataset.py:
```diff
@@ -0,0 +1,150 @@
+def convert_to_rgb(image):
+    return image.convert("RGB")
+
+def get_transform(image_size=384):
```
这些函数不需要吧？

cuicheng01 commented on ppcls/data/dataloader/ram_dataset.py:
```diff
@@ -0,0 +1,150 @@
+
+
+
+class RAM_pretrain_dataset(Dataset):
```
建议改为RAMPretrainDataset，下同

cuicheng01 commented on ppcls/data/postprocess/ramoutput.py:
```diff
@@ -0,0 +1,63 @@
+import os
```
版权信息和引用信息

cuicheng01 commented on ppcls/engine/engine_multimodal.py:
```diff
@@ -0,0 +1,142 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
2023

cuicheng01 commented on ppcls/engine/engine_multimodal.py:
```diff
@@ -0,0 +1,142 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import absolute_import
```
future可以去掉了

cuicheng01 commented on ppcls/engine/train/train_multimodal.py:
```diff
@@ -0,0 +1,96 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
2023

cuicheng01 commented on ppcls/engine/train/train_multimodal.py:
```diff
@@ -0,0 +1,96 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import absolute_import, division, print_function
```
去掉

cuicheng01 commented on ppcls/loss/cliploss.py:
```diff
@@ -0,0 +1,22 @@
+import paddle
```
版权信息

cuicheng01 commented on ppcls/loss/ramloss.py:
```diff
@@ -0,0 +1,25 @@
+        super().__init__()
+        self.mode = mode
+
+    ## **kwargs 参数没有实际意义，仅作为兼容框架本身所使用的参数
```
注释写到函数下边，英文注释

cuicheng01 commented on tools/infer_multimodal.py:
```diff
@@ -0,0 +1,18 @@
+from __future__ import absolute_import
```
去掉，增加版权信息

cuicheng01 commented on tools/infer_multimodal.py:
```diff
@@ -0,0 +1,18 @@
+from __future__ import print_function
+import os
+import sys
+__dir__ = os.path.dirname(os.path.abspath(__file__))
```
注意规范和顺序

cuicheng01 commented on tools/train_multimodal.py:
```diff
@@ -0,0 +1,34 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
2023

cuicheng01 commented on tools/train_multimodal.py:
```diff
@@ -0,0 +1,34 @@
+from __future__ import print_function
+import os
+import sys
+__dir__ = os.path.dirname(os.path.abspath(__file__))
```
注意规范和顺序

cuicheng01 commented on tools/train_multimodal.py:
```diff
@@ -0,0 +1,34 @@
+
+if __name__ == "__main__":
+    args = config.parse_args()
+    args.config = "./ppcls/configs/ram/config_train.yaml"
```
这不要写死

cuicheng01 reviewed (commented):
1.comment需要好好改一下，所有的py需要注意下代码规范
2.建议ram放到arch下，不要放在backbone下
3.所有的文件需要过pre-commit
4.需要增加详细的文档，包括训练，infer、推理，可以参考：https://github.com/PaddlePaddle/PaddleClas/blob/develop/ppcls/configs/MultiLabelCOCO/MLDecoder/README.md

psky1111 committed:
pre-commit format fix

psky1111 committed:
update ram doc

psky1111 committed:
update

psky1111 commented:
updated ram parameters：
链接：https://pan.baidu.com/s/1UKn5yfRsV6KIP4K5rgK8VA 
提取码：dbvi 
--来自百度网盘超级会员V6的分享

psky1111 committed:
fix format problem

psky1111 committed:
pre-commit fix

psky1111 committed:
update

psky1111 committed:
pre-commit

psky1111 committed:
update format

psky1111 committed:
Update __init__.py

psky1111 commented:
clip pretrain param:
链接：https://pan.baidu.com/s/1Nx0quYYkpNOlmkkv5mt8mg 
提取码：lx88 
--来自百度网盘超级会员V6的分享
clip pretrained param script:
"""python
map_key = {
    "in_proj_": "in_proj."
}
def paddle_to_torch(torch_dict, paddle_dict):
    new_paddle_dict = {}
    for torch_key in torch_dict.keys():
        paddle_key = torch_key
        for mm_k in map_key.keys():
            if mm_k in paddle_key:
                paddle_key = paddle_key.replace(mm_k, map_key[mm_k])
        if ('out_proj.weight' in paddle_key) or ("in_proj.weight" in paddle_key) or  (("mlp" in paddle_key) and ("weight") in paddle_key):
            paddle_tensor = paddle.to_tensor(torch_dict[torch_key].cpu().numpy().transpose())
            shape_ori, shape_new = paddle_dict[paddle_key].shape, paddle_tensor.shape
            if shape_ori == shape_new:
                new_paddle_dict[paddle_key] = paddle_tensor
            else:
                new_paddle_dict[paddle_key] = paddle_tensor.T
        else:
            paddle_tensor = paddle.to_tensor(torch_dict[torch_key].cpu().numpy())
            shape_ori, shape_new = paddle_dict[paddle_key].shape, paddle_tensor.shape
            if shape_ori == shape_new:
                new_paddle_dict[paddle_key] = paddle_tensor
            else:
                new_paddle_dict[paddle_key] = paddle_tensor.T
    return new_paddle_dict
"""

psky1111 commented:
ram code with readme
链接：https://pan.baidu.com/s/1X1qvQs8z3zOk8VTQI9g4Pw 
提取码：bi58 
--来自百度网盘超级会员V6的分享

psky1111 committed:
update

cuicheng01 commented on .gitignore:
```diff
@@ -15,3 +15,4 @@ nohup.out
 .idea
 inference/
 test.py
+clip_text.py
```
这个去掉吧

cuicheng01 commented on deploy/configs/inference_ram.yaml:
```diff
@@ -0,0 +1,38 @@
+        size: 384
+    - NormalizeImage:
+        scale: 0.00392157
+        mean: [0.485, 0.456, 0.406]
```
官方的CLIP的mean和std不是这个值，这里需要确认下要不要改

cuicheng01 commented on deploy/python/postprocess.py:
```diff
@@ -71,7 +71,7 @@ def __init__(self, func_list, main_indicator="Topk"):
         rtn = None
         for func in self.func_list:
-            tmp = func(x, image_file)
+            tmp = func(*x, image_file)
```
改这个的目的是？

cuicheng01 commented on deploy/python/postprocess.py:
```diff
@@ -496,3 +496,65 @@ def __call__(self, batch_preds, file_names=None):
+                 delete_tag_index=[],
+                 ram_class_threshold_path=""):
+        self.language = language
+        assert tag_list, tag_list_chinese
```
这里需要给出报错信息

cuicheng01 commented on deploy/python/predict_multimodal.py:
```diff
@@ -0,0 +1,67 @@
+import cv2
+import numpy as np
+
+from paddleclas.deploy.utils import logger, config
```
这里不需要从paddleclas下到入吧？

cuicheng01 commented on ppcls/arch/clip/clip.py:
```diff
@@ -0,0 +1,595 @@
+import paddle.nn as nn
+from paddle.nn import functional as F
+from paddle.nn.initializer import Assign, Normal, Constant
+from paddle.vision.transforms import Compose, Resize, CenterCrop, ToTensor, Normalize
```
这些操作不需要了吧？

cuicheng01 commented on ppcls/arch/clip/clip.py:
```diff
@@ -0,0 +1,595 @@
+        return results
+
+
+def get_transforms(image_resolution):
```
这些可以去掉了吧？

cuicheng01 commented on ppcls/arch/clip/clip.py:
```diff
@@ -0,0 +1,595 @@
+        transformer_width=768,
+        transformer_heads=12,
+        transformer_layers=12)
+    return model, get_transforms(224)
```
get_transforms()的逻辑统一在yaml中指定

cuicheng01 commented on ppcls/arch/clip/clip.py:
```diff
@@ -0,0 +1,595 @@
+    return model, get_transforms(224)
+
+
+CLIP_DICT = {
```
这个没有存在的必要了吧？

cuicheng01 commented on ppcls/arch/ram/ram_plus.py:
```diff
@@ -0,0 +1,250 @@
+
+
+# load RAM pretrained model parameters
+def ram_plus(pretrained='', **kwargs):
```
这里的pretrained下边没有使用？

cuicheng01 commented on ppcls/configs/ram/RAM.yaml:
```diff
@@ -0,0 +1,170 @@
+            backend: pil
+        - NormalizeImage:
+            scale: 1.0/255.0
+            mean: [0.485, 0.456, 0.406]
```
mean和std需要确认下

cuicheng01 commented on ppcls/configs/ram/RAM.yaml:
```diff
@@ -0,0 +1,170 @@
+  Train:
+    dataset:
+      name: RAMPretrainDataset
+      ann_file: [./visual-genome/vg_ram.json]
```
文档中需要注明vg_ram.json的数据集长啥样子

cuicheng01 commented on ppcls/configs/ram/RAM_plus.yaml:
```diff
@@ -0,0 +1,173 @@
+            backend: pil
+        - NormalizeImage:
+            scale: 1.0/255.0
+            mean: [0.485, 0.456, 0.406]
```
mean和std需要确认下

cuicheng01 commented on ppcls/configs/ram/README.md:
```diff
@@ -0,0 +1,394 @@
+
+| Model | BackBone   | Size   | Inference Prompt | OpenImages-MAP |
+|-------|------------|--------|------------------|----------------|
+| RAM   | Swin-large | 5.63GB | LLM Tag Dec      | 82.2           |
```
这里需要指明CLIP用的是哪个模型

cuicheng01 commented on ppcls/configs/ram/README.md:
````diff
@@ -0,0 +1,394 @@
+* 前往官方[repo](https://github.com/xinyu1205/recognize-anything/tree/main)下载对应数据集json文件。同时按照json文件目录格式，准备相应的数据。目录格式为：
+```json
+{
+    {
````
里边的每个字段需要解释下

cuicheng01 commented on ppcls/configs/ram/README.md:
````diff
@@ -0,0 +1,394 @@
+```
+**注意:**
+1. 目前多标签分类的损失函数默认使用`AsymmetricLoss`。
+2. 目前多标签分类的评估指标默认使用`MAP(integral)`。
````
注意的这两个不太对吧？

cuicheng01 commented on ppcls/configs/ram/README.md:
````diff
@@ -0,0 +1,394 @@
+
+得到类似下面的输出：
+```
+{'class_ids': [[0, 593], [0, 871], [0, 998], [0, 2071], [0, 3336], [0, 3862]], 'scores': [871], 'label_names': ['棕色 | 鸡 | 公鸡 | 母鸡  | 红色 | 站/矗立/摊位']}
````
这个scores是？

cuicheng01 commented on ppcls/configs/ram/README.md:
```diff
@@ -0,0 +1,394 @@
+python3 -m paddle.distributed.launch \
+    --gpus="0,1,2,3" \
+    tools/train_multimodal.py \
+        -c ./ppcls/configs/ram/RAM.yaml
```
只有训练，没有finetune示例？

cuicheng01 commented on ppcls/configs/ram/RAM_plus.yaml:
```diff
@@ -0,0 +1,173 @@
+
+    sampler:
+      name: DistributedBatchSampler
+      batch_size: 52
```
单卡bs

psky1111 committed:
update

psky1111 committed:
Update .gitignore

psky1111 commented:
[models--bert-base-uncased.zip](https://github.com/PaddlePaddle/PaddleClas/files/13848891/models--bert-base-uncased.zip)

psky1111 committed:
fix bug

1. the behavior of static inference is different and thus need to reshape dimension.
2. training dataloader add collect_fn to replace original one. 3. fix the threshold.

psky1111 committed:
fix train gpu problem

psky1111 committed:
compatiable for static graph

psky1111 committed:
update format

psky1111 committed:
fix format problem

psky1111 committed:
del the static support

psky1111 committed:
update

1. fix ram training problem.
2. finish clip vetorlize inference

psky1111 committed:
Update ram_dataset.py

psky1111 committed:
add print

psky1111 committed:
remove clip part

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/PaddlePaddle/PaddleClas?pullRequest=3062) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/PaddlePaddle/PaddleClas?pullRequest=3062) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/PaddlePaddle/PaddleClas?pullRequest=3062) it.</sub>
